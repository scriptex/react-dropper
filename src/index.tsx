import * as React from 'react';

interface IDropperProps {
	size: number;
	width: number;
	height: number;
	image: any;
	onSelect: (args: any) => any;
	onMouseMove: (args: any) => any;
	onTouchMove: (args: any) => any; // TODO support touch events
}

interface IDropperState {
	offsetTop: number;
	offsetLeft: number;
}

export class Dropper extends React.Component<IDropperProps, IDropperState> {
	private canvas: React.Ref<HTMLCanvasElement> = React.createRef();
	private ctx: CanvasRenderingContext2D;

	public state: IDropperState = {
		offsetTop: 0,
		offsetLeft: 0
	};

	public componentDidMount(): void {
		const { current } = this.canvas as any;
		const offset = current.getBoundingClientRect();

		this.ctx = current.getContext('2d');

		this.setState({
			offsetTop: offset.top,
			offsetLeft: offset.left
		});
	}

	public render(): React.ReactNode {
		this.drawImage();

		return (
			<canvas
				className="tile__colorpicker"
				width={this.props.width}
				height={this.props.height}
				ref={this.canvas}
				onClick={this.setColor}
				onMouseMove={this.setColor}
			/>
		);
	}

	private drawImage = () => {
		const { width, height, image } = this.props;
		const htmlImage: HTMLImageElement = new Image();

		htmlImage.onload = () => {
			this.ctx.drawImage(htmlImage, 0, 0, width, height);
		};

		htmlImage.src = image as any;
	};

	private getImageData = (e: React.MouseEvent<HTMLCanvasElement>): Uint8ClampedArray => {
		const canvasX = Math.floor(e.pageX - this.state.offsetLeft);
		const canvasY = Math.floor(e.pageY - this.state.offsetTop);
		const imageData = this.ctx.getImageData(canvasX, canvasY, 1, 1);

		return imageData.data;
	};

	private setColor = (e: React.MouseEvent<HTMLCanvasElement>): void => {
		e.preventDefault();

		const [r, g, b]: Uint8ClampedArray = this.getImageData(e);
		const color: string = `#${(b + 256 * g + 65536 * r).toString(16)}`;

		this.props.onSelect({
			...this.state,
			color
		});
	};
}

export default Dropper;
