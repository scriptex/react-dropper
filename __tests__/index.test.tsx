import * as React from 'react';
import { act, render, waitFor, fireEvent, RenderResult } from '@testing-library/react';

import { Dropper } from '../src';

const snapshotTest = (
  Component: React.FC<Readonly<React.ComponentProps<any>>>,
  callback?: (result: RenderResult) => void
): void => {
  const { name } = Component;

  describe(`${name} component`, () => {
    it(`Should render the ${name} component`, async () => {
      let result!: RenderResult;

      await act(async () => {
        result = await waitFor(() => render(<Component />));
      });

      const { asFragment } = result;

      expect(asFragment()).toMatchSnapshot();

      if (typeof callback === 'function') {
        callback(result);
      }
    });

    it(`Should unmount the ${name} component`, async () => {
      let result: RenderResult;

      await act(async () => {
        result = await waitFor(() => render(<Component />));
      });

      // @ts-ignore
      const { unmount, asFragment } = result;

      unmount();

      expect(asFragment()).toMatchSnapshot();
    });
  });
};

snapshotTest(Dropper, (result: RenderResult) => {
  const { container, asFragment } = result;

  fireEvent.click(container.querySelector('canvas')!, {
    pageX: 300,
    pageY: 200,
    preventDefault: jest.fn()
  });

  expect(asFragment()).toMatchSnapshot();

  fireEvent.mouseMove(container.querySelector('canvas')!, {
    pageX: 300,
    pageY: 200,
    preventDefault: jest.fn()
  });

  expect(asFragment()).toMatchSnapshot();
});
