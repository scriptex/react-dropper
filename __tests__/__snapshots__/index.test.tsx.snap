// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropper component Should render the Dropper component 1`] = `
<DocumentFragment>
  <canvas
    class="react-dropper"
    height="150"
    width="300"
  />
</DocumentFragment>
`;

exports[`Dropper component Should render the Dropper component 2`] = `
<DocumentFragment>
  <canvas
    class="react-dropper"
    height="150"
    width="300"
  />
</DocumentFragment>
`;

exports[`Dropper component Should render the Dropper component 3`] = `
<DocumentFragment>
  <canvas
    class="react-dropper"
    height="150"
    width="300"
  />
</DocumentFragment>
`;

exports[`Dropper component Should unmount the Dropper component 1`] = `<DocumentFragment />`;
